File: backend.scala/app/com/skewwhiffy/auraltester/internalnotation/IntervalFactory.scala
package com.skewwhiffy.auraltester.internalnotation

import com.skewwhiffy.auraltester.notes.{DirectedInterval, Interval}

import scala.util.chaining.scalaUtilChainingOps

class IntervalFactory {
  def getDirectedIntervals(rawIntervals: String): List[DirectedInterval] = rawIntervals
    .split(' ')
    .map(getDirectedInterval)
    .toList

  def getDirectedInterval(rawInterval: String): DirectedInterval = {
    getRawDeviations(rawInterval)
      .toCharArray
      .foldRight(getBaseInterval(rawInterval))((it, current) => it match {
        case '-' => current.diminished
        case '+' => current.augmented
        case _ => throw new IllegalArgumentException(s"Not a valid deviation: '$it'")
      })
      .pipe(it => {
        getRawDirection(rawInterval) match {
          case "-" => it.down
          case "+" | "" => it.up
          case it => throw new IllegalArgumentException(s"Not a valid direction: '$it''")
        }
      })
  }

  private def getBaseInterval(rawInterval: String): Interval = getRawInt(rawInterval).toInt.pipe {
    case it@(1 | 4 | 5 | 8) => Interval.perfect(it)
    case it => Interval.major(it)
  }

  private def getRawDirection(rawInterval: String) = rawInterval match {
    case it if it.startsWith("+") => "+"
    case it if it.startsWith("-") => "-"
    case _ => ""
  }

  private def getRawInt(rawInterval: String) = {
    def getRawInt(soFar: String, remaining: String): String = {
      try {
        remaining
          .substring(0, 1)
          .toInt
          .pipe(it => s"$soFar$it")
          .pipe(it => getRawInt(it, remaining.substring(1)))
      } catch {
        case _: NumberFormatException => soFar
        case _: StringIndexOutOfBoundsException => soFar
      }
    }

    if (rawInterval.startsWith("+") || rawInterval.startsWith("-")) getRawInt("", rawInterval.substring(1))
    else getRawInt("", rawInterval)
  }

  private def getRawDeviations(rawInterval: String): String = rawInterval
    .substring(getRawInt(rawInterval).length + getRawDirection(rawInterval).length)
}

File: backend/src/main/kotlin/com/skewwhiffy/notation/factory/IntervalFactory.kt
package com.skewwhiffy.notation.factory

import com.skewwhiffy.notation.model.interval.DirectedInterval
import com.skewwhiffy.notation.model.interval.Interval
import org.springframework.stereotype.Service

@Service
class IntervalFactory {
  fun getDirectedIntervals(rawIntervals: String) = rawIntervals
    .split(' ')
    .map(::getDirectedInterval)
    .toList()

  fun getDirectedInterval(rawInterval: String): DirectedInterval =
    getRawDeviations(rawInterval)
      .toCharArray()
      .fold(getBaseInterval(rawInterval)) { current, it ->
        when (it) {
          '-' -> current.diminished
          '+' -> current.augmented
          else -> throw IllegalArgumentException("Not a valid deviation: '$it'")
        }
      }
      .let {
        when (getRawDirection(rawInterval)) {
          "-" -> it.down
          "+" -> it.up
          "" -> it.up
          else -> throw IllegalArgumentException("Not a valid direction: '$it''")
        }
      }

  private fun getBaseInterval(rawInterval: String): Interval = getRawInt(rawInterval)
    .toInt()
    .let { if (listOf(1, 4, 5, 8).contains(it)) Interval.perfect(it) else Interval.major(it) }

  private fun getRawDirection(rawInterval: String): String = "+-"
    .toCharArray()
    .find { rawInterval.startsWith(it) }
    ?.toString() ?: ""

  private fun getRawInt(rawInterval: String): String {
    fun getRawInt(soFar: String, remaining: String): String = try {
      remaining
        .substring(0, 1)
        .toInt()
        .let { "$soFar$it" }
        .let { getRawInt(it, remaining.substring(1)) }
    } catch (ex: NumberFormatException) {
      soFar
    } catch (ex: StringIndexOutOfBoundsException) {
      soFar
    }

    return if (rawInterval.startsWith("+") || rawInterval.startsWith("-")) getRawInt("", rawInterval.substring(1))
    else getRawInt("", rawInterval)
  }

  private fun getRawDeviations(rawInterval: String): String = rawInterval
    .substring(getRawInt(rawInterval).length + getRawDirection(rawInterval).length)
}